contracts/zaps/ZapCurveTokenToPrincipalToken.sol

```diff
@@ -0,0 +1,458 @@
+    {
+        // All curvePools have either 2 or 3 "root" tokens
+        require(
+            _zap.amounts.length == 2 || _zap.amounts.length == 3,
+            "!(2 >= amounts.length <= 3)"
+        );
+
+        // Flag if the current zap has amounts. Used to short-circuit
+        // unnecessary calls
+        bool zapHasAmounts = false;
+        // Flag for where the user has a childZap with amounts. Used to
+        // short-circuit unnecessary calls
+        bool ctxHasAmounts = false;
+        for (uint8 i = 0; i < _zap.amounts.length; i++) {
+            // Must check we do not unintentionally send ETH
+            if (_zap.roots[i] == _ETH_CONSTANT)
+                require(msg.value == _zap.amounts[i], "incorrect value");
+
+            // Setting ctxHasAmounts flag as true means that there
+            // was a previous childZap we must acknowledge
+            if (!ctxHasAmounts && _ctx[i] > 0) {
+                ctxHasAmounts = true;
+            }
+
+            if (_zap.amounts[i] > 0) {
+                // If the zapHasAmounts has not been set, we set it to true
+                if (!zapHasAmounts) {
+                    zapHasAmounts = true;
+                }
+
+                if (_zap.roots[i] == _ETH_CONSTANT) {
+                    // We build the context container with our amounts
+                    _ctx[i] += _zap.amounts[i];
+                } else {
+                    // In the case of swapping an ERC20 "root" we must transfer them
+                    // to this contract in order to make the exchange
+                    IERC20(_zap.roots[i]).safeTransferFrom(
+                        msg.sender,
+                        address(this),
+                        _zap.amounts[i]
+                    );
+
+                    // Due to rounding issues of some tokens, we find the
+                    // relevant token balance of this contract
+                    _ctx[i] = IERC20(_zap.roots[i]).balanceOf(address(this));
+                }
+            }
+        }
+
+        // When we do not have anything to swap for, we short-circuit
+        if (!zapHasAmounts && !ctxHasAmounts) {
+            return 0;
+        }
+
+        // It is necessary to add liquidity to the respective curve pool like
+        // this due to the non-standard interface of the function in the curve
+        // contracts. Not only is there two variants of fixed length array
+        // amount inputs but it is often inconsistent whether return values
+        // exist or not
+        // Also, the context amounts container, although fixed-length of 2 or 3,
+        // because it is a low-level function call, solidity will not complain
+        // and the first 2 indexes are only considered in the case of it being
+        // length 2
+        address(_zap.curvePool).functionCallWithValue(
+            abi.encodeWithSelector(_zap.funcSig, _ctx, 0),
+            msg.value
+        );
+
+        return _zap.lpToken.balanceOf(address(this));
```

Comment:
The reason for this was that in some cases, I think in particular the `STETH` token, there is a precision loss of 1 wei. If by the differential value, you mean the `zap.amounts[i]` figure, that would cause a revert in the curve call